tsdb/head.go

```diff
@@ -1663,7 +1663,7 @@ func (h *Head) getOrCreate(hash uint64, lset labels.Labels) (*memSeries, bool, e
 
 func (h *Head) getOrCreateWithID(id chunks.HeadSeriesRef, hash uint64, lset labels.Labels) (*memSeries, bool, error) {
 	s, created, err := h.series.getOrSet(hash, lset, func() *memSeries {
-		return newMemSeries(lset, id, h.opts.IsolationDisabled)
+		return newMemSeries(lset, id, labels.StableHash(lset), h.opts.IsolationDisabled)
```

Comment:
Some extra cost here to do a second hash, but only once for each series created.
We could adjust `StableHash` to take the hash we already have here and return that if it is known to be the correct value.
(Which it will be for labels-as-slice and the upcoming implementation in #12304)

Reply:
WDYT about making it conditional, adding a `EnableSharding` config option to TSDB, like proposed in https://github.com/prometheus/prometheus/pull/10421/commits/c38d4e494c117d1700fddd0d32cb78309ccce1e1 ?